### PR TITLE
[NTRN-163] add dex stargate helpers

### PR DESCRIPTION
**Pull request by NeverHappened**
**task:** https://hadronlabs.atlassian.net/browse/NTRN-163

**this PR:**
- removes `build_proto` stuff the repo contains for it's outdated and actually useless;
- adds a `neutron_sdk::stargate::proto_types` package which contains all neutron-specific proto types generated in rust;
- adds a `neutron_sdk::stargate` package which contains stargate query/msg helper funcs and types only for the dex module atm.

**to be done in by a future release:**
- add stargate helpers for the rest of the neutron modules;
- add proto generator script (partially done by @sotnikov-s, but not in scope of this PR because of time pressure);
- fix dex module imperfections (unix timestamp as `i64` instead of `Timestamp`, `Pool_metadata` -> `pool_metadata`);
- implement `PrecDec` data type (used here https://github.com/neutron-org/neutron/blob/168815d3703a803f97566f2ddaf0954ac8bda203/proto/neutron/dex/params.proto#L14 and in a couple of other places).

**related PRs:**
- https://github.com/neutron-org/neutron-dev-contracts/pull/37
- https://github.com/neutron-org/neutron-integration-tests/pull/251
- https://github.com/neutron-org/neutronjsplus/pull/12

### CODE REVIEW

**Comment by sotnikov-s** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/7111070219

File: packages/neutron-sdk/src/stargate/aux.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    Res: DeserializeOwned,
+{
+    deps.querier.query(&QueryRequest::Stargate {
+        path: path.to_string(),
+        data: req.encode_to_vec().into(),
+    })
+}
+
+pub(crate) fn create_stargate_msg<Req: prost::Message>(req: Req, path: &str) -> CosmosMsg {
+    cosmwasm_std::CosmosMsg::Stargate {
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/aux.rs:17]:
```suggestion
    CosmosMsg::Stargate {
```

File: packages/neutron-sdk/src/stargate/types_dex.rs [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+    }
+}
+
+/// deserialize_expiration_time deserealizes an optional expiration_time value on dex module's rules:
+/// - if it's None, it returns None (non-expiring limit orders);
+/// - if it's a default golang time.Time value used for JIT limit order type (0001-01-01T00:00:00Z),
+/// it returns 0, because the timestamp for this value is invalid (-62135596800);
+/// - in the rest of the cases, it assumes it's a valid RFC 3339 formatted date time and tries to
+/// parse it and returns a unix timestamp.
+fn deserialize_expiration_time<'de, D>(deserializer: D) -> Result<Option<Int64>, D::Error>
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/types_dex.rs:774]: It's unused?

File: packages/neutron-sdk/src/stargate/aux.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+use cosmwasm_std::{Binary, CosmosMsg, Deps, QueryRequest, StdResult};
+use prost_types::Timestamp as TimestampGen;
+use serde::de::DeserializeOwned;
+
+pub(crate) fn make_stargate_query<Req, Res>(deps: Deps, req: Req, path: &str) -> StdResult<Res>
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/aux.rs:5]: I'd add a doc comment for the func, that describes a purpose of the function, arguments and a returned value

File: packages/neutron-sdk/src/stargate/aux.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    Req: prost::Message,
+    Res: DeserializeOwned,
+{
+    deps.querier.query(&QueryRequest::Stargate {
+        path: path.to_string(),
+        data: req.encode_to_vec().into(),
+    })
+}
+
+pub(crate) fn create_stargate_msg<Req: prost::Message>(req: Req, path: &str) -> CosmosMsg {
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/aux.rs:16]: I'd add a doc comment for the func, that describes a purpose of the function, arguments and a returned value

File: packages/neutron-sdk/src/stargate/aux.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    cosmwasm_std::CosmosMsg::Stargate {
+        type_url: path.to_string(),
+        value: Binary::from(req.encode_to_vec()),
+    }
+}
+
+pub(crate) fn convert_timestamp(timestamp: i64) -> TimestampGen {
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/aux.rs:23]: I think name of the method is not very self-describable.
Maybe rename it to smth `proto_timestamp_from_i64`?

File: packages/neutron-sdk/src/stargate/msg_dex.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+use cosmwasm_std::CosmosMsg;
+
+const DEPOSIT_MSG_PATH: &str = "/neutron.dex.MsgDeposit";
+const WITHDRAWAL_MSG_PATH: &str = "/neutron.dex.MsgWithdrawal";
+const PLACE_LIMIT_ORDER_MSG_PATH: &str = "/neutron.dex.MsgPlaceLimitOrder";
+const WITHDRAW_FILLED_LIMIT_ORDER_MSG_PATH: &str = "/neutron.dex.MsgWithdrawFilledLimitOrder";
+const CANCEL_LIMIT_ORDER_MSG_PATH: &str = "/neutron.dex.MsgCancelLimitOrder";
+const MULTI_HOP_SWAP_MSG_PATH: &str = "/neutron.dex.MsgMultiHopSwap";
+
+pub fn msg_deposit(req: DepositRequest) -> CosmosMsg {
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/msg_dex.rs:19]: What do you think of adding doc comments for each method which describes what does it do?

File: packages/neutron-sdk/src/stargate/query_dex.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+const POOL_RESERVES_ALL_QUERY_PATH: &str = "/neutron.dex.Query/PoolReservesAll";
+const POOL_RESERVES_QUERY_PATH: &str = "/neutron.dex.Query/PoolReserves";
+const ESTIMATE_MULTI_HOP_SWAP_QUERY_PATH: &str = "/neutron.dex.Query/EstimateMultiHopSwap";
+const ESTIMATE_PLACE_LIMIT_ORDER_QUERY_PATH: &str = "/neutron.dex.Query/EstimatePlaceLimitOrder";
+const POOL_QUERY_PATH: &str = "/neutron.dex.Query/Pool";
+const POOL_BY_ID_QUERY_PATH: &str = "/neutron.dex.Query/PoolByID";
+const POOL_METADATA_QUERY_PATH: &str = "/neutron.dex.Query/PoolMetadata";
+const POOL_METADATA_ALL_QUERY_PATH: &str = "/neutron.dex.Query/PoolMetadataAll";
+
+pub fn get_params(deps: Deps, req: ParamsRequest) -> StdResult<ParamsResponse> {
```
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/query_dex.rs:49]: What do you think of adding doc comments for each method which describes what does it do?

**Review by pr0n00gler — COMMENTED** [review]
> add proto generator script (partially done by @sotnikov-s, but not in scope of this PR because of time pressure);

Could you attach at least a WIP script here?
Because we need to finish it at least

File: packages/neutron-sdk/src/stargate/msg_dex.rs
Comment (pr0n00gler) [inline packages/neutron-sdk/src/stargate/msg_dex.rs:1]: What if we move all related dex files (msg_dex, query_dex, types_dex) into separate dex folder?
So the final directory structure would look like:
* stargate/
    * dex/
        * msg.rs
        * query.rs
        * types.rs

**Review by swelf19 — APPROVED** [review]
LGTM.
And a suggestion, i can see there is the `Name` trait in prost repo - https://github.com/tokio-rs/prost/blob/master/src/name.rs#L7 which potentially could let you avoid handcrafted helpers (msg_dex, query_dex) by generating appropriate type names.

File: packages/neutron-sdk/src/stargate/types_dex.rs (hunk 2 above)
Comment (swelf19) [reply packages/neutron-sdk/src/stargate/types_dex.rs:774]: it's used in serde macros at line 688

File: packages/neutron-sdk/src/stargate/msg_dex.rs
Comment (sotnikov-s) [reply packages/neutron-sdk/src/stargate/msg_dex.rs:1]: sounds good. I also find the proto_types directory inappropriate. I'll rethink folders design and probably move it out of the stargate package

**Review by pr0n00gler — APPROVED** [review]

**Comment by sotnikov-s** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/7111070219

**Comment by sotnikov-s** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/7206923966

**Pull request merged**